### PR TITLE
Add rbac logic and FQDN template field

### PR DESCRIPTION
**Pull request by alebedev87**
What's inside:
- RBAC (cluster role, cluster role binding) logic is added with unit tests
- Kubebuilder's RBAC annotation for operands
- `FQDNTemplate` API field added as without it ExternalDNS doesn't start when `HostnameAnnotationPolicy` is set to `Ignore` (default). Functionally it's needed to give FQDN names to services because the other possibility (via a special annotation) is disabled by `--ignore-hostname-annotation` flag (comes from the previously mentioned Ignore policy):
```bash
  --ignore-hostname-annotation   Ignore hostname annotation when generating DNS names, valid only when using fqdn-template is set
```
- Re-added tolerations for ExternalDNS deployment (somehow I lost them along the way)

### CODE REVIEW

**Comment by alebedev87** [comment]
- Put vendored files into a separate commit
- Added first remarks after the discussion with @frobware

File: README.md
Comment (alebedev87) [reply README.md:24]: Yes, I put it here just as a reminder before Luigi and I made all these things official Jira user stories.

File: pkg/operator/controller/externaldns/rbac.go
Comment (alebedev87) [reply pkg/operator/controller/externaldns/rbac.go:239]: It's a binding of the cluster role to the ExternalDNS deployment's service account (Subject).
I don't expect any other service account or user to get the externalDNS role in the near future.

File: pkg/operator/controller/externaldns/rbac.go
Comment (alebedev87) [reply pkg/operator/controller/externaldns/rbac.go:194]: Yes, indeed, the pattern is different: not very consistent, I agree but the cases are slightly different.
In case of the cluster role I reset all the rules no matter the change - to avoid the complex resetting of `updated` variable, so it's a single field.
In the cluster role binding case `updated` parameter can be set with the desired values in 3 different fields.

I changed the comment of `externalDNSRoleBindingChanged` to better clarify this.

File: pkg/operator/controller/externaldns/rbac.go
Comment (alebedev87) [reply pkg/operator/controller/externaldns/rbac.go:34]: updated the comment

File: pkg/operator/controller/externaldns/rbac.go
Comment (alebedev87) [reply pkg/operator/controller/externaldns/rbac.go:54]: I think that returning the current state of the object is consistent with the first (`bool`) parameter: we say that the role indeed exists here its value but the update failed.

Updated the comment to clarify this.

File: config/rbac/role.yaml
Comment (alebedev87) [reply config/rbac/role.yaml:12]: Oouf, it took me some time to find out all about EndpointSlices and dualstack support.

1. This cluster role has the same resources as [the upstream one](https://github.com/kubernetes-sigs/external-dns/blob/master/kustomize/external-dns-clusterrole.yaml#L7).
2. `EndpointSlice` is not used anywhere in the code of ExternalDNS
3. No `AAAA` record is created for dualstack sources:
```bash
# ClusterIP source service
$ oc get svc httpserver2 -o yaml | grep -A 14 '^spec:'
spec:
  clusterIP: 172.30.46.13
  clusterIPs:
  - 172.30.46.13
  - fd02::9bfb
  ipFamilies:
  - IPv4
  - IPv6
  ipFamilyPolicy: PreferDualStack
  ports:
  - port: 8080
    protocol: TCP
    targetPort: 8080
  selector:
    run: httpserver

# externalDNS logs    
$ oc -n external-dns logs deploy/external-dns-sample | grep 'publish-external-dns/httpserver2'
time="2021-09-16T15:00:43Z" level=debug msg="Endpoints generated from service: publish-external-dns/httpserver2: [httpserver2.external-dns-test.alebedev.com 0 IN A  172.30.46.13 []]"
time="2021-09-16T15:01:43Z" level=debug msg="Endpoints generated from service: publish-external-dns/httpserver2: [httpserver2.external-dns-test.alebedev.com 0 IN A  172.30.46.13 []]"

# A record which returns IPv4 ClusterIP
aws route53 list-resource-record-sets --output json --hosted-zone-id "/hostedzone/Z06076472WHIZX9L24NVR" --query "ResourceRecordSets[?Type == 'A']"
[
    {
        "Name": "httpserver2.external-dns-test.alebedev.com.",
        "Type": "A",
        "TTL": 300,
        "ResourceRecords": [
            {
                "Value": "172.30.46.13"
            }
        ]
    }
]
```

```bash
# LoadBalancer service source
$ oc get svc httpserver -o yaml | grep -A 25 '^spec:'
spec:
  clusterIP: 172.30.51.48
  clusterIPs:
  - 172.30.51.48
  - fd02::30c0
  externalTrafficPolicy: Cluster
  ipFamilies:
  - IPv4
  - IPv6
  ipFamilyPolicy: PreferDualStack
  ports:
  - nodePort: 31189
    port: 8080
    protocol: TCP
    targetPort: 8080
  selector:
    run: httpserver
  sessionAffinity: None
  type: LoadBalancer
status:
  loadBalancer:
    ingress:
    - hostname: abad68df9bcad4b7387d8a66d8fc2758-743607947.us-east-2.elb.amazonaws.com

# A record which redirects to AWS ELB DNS name (from Service's status)    
$ aws route53 list-resource-record-sets --output json --hosted-zone-id "/hostedzone/Z06076472WHIZX9L24NVR" --query "ResourceRecordSets[?Type == 'A']"
[
    {
        "Name": "httpserver.external-dns-test.alebedev.com.",
        "Type": "A",
        "AliasTarget": {
            "HostedZoneId": "Z3AADJGX6KTTL2",
            "DNSName": "abad68df9bcad4b7387d8a66d8fc2758-743607947.us-east-2.elb.amazonaws.com.",
            "EvaluateTargetHealth": true
        }
    },
]
```
4. For the moment the dualstack support is limited to the Ingress source which is backed-up by [ALB](https://github.com/kubernetes-sigs/external-dns/blob/43d29f8032d2e79d2d666f794b89b71ad3349a12/docs/tutorials/alb-ingress.md#dualstack-albs) or [Zalando's AWS](https://github.com/kubernetes-sigs/external-dns/blob/7b8c69b22e33e04041c3a8299e13cab4f44ad486/docs/tutorials/kube-ingress-aws.md#dualstack-albs) ingress controller. For such ingress objects it's possible to add `alb.ingress.kubernetes.io/ip-address-type: dualstack` annotation which will instruct the externalDNS to create `AAAA` record which will target tthe IP from the status of the ingress resources. Note: the status is populated by the controller with IPv4, IPv6 addresses of the LB created in cloud.

So, I don't think ExternalDNS does anything with `EndpointSlice` resource because even for the real dualstack cases it doesn't go that deep in Kubernetes network stack. As a matter of fact I'm not even sure that ExternalDNS needs rights for the `Endpoints` resource :) but I'd rather keep it as it's used in the upstream.

Overall IPv6 support is not very consolidated - mostly left into consideration of the provider. There is this issue though which tries to Xref all the effort: [2051](https://github.com/kubernetes-sigs/external-dns/issues/2051).

**Comment by alebedev87** [comment]
@frobware :
- Test for `Reconcile` function added. The fake controller-runtime client is used.
- Talking about a dummy provider, I think that it's would be more suitable for an e2e test to test the link between the operator and the DNS provider. We have planned a dedicated user story for the e2e tests.

**Comment by alebedev87** [comment]
Reconcile test modified to better check the expected events.

File: config/rbac/role.yaml
Comment (alebedev87) [reply config/rbac/role.yaml:12]: ExternalDNS seems to take what is in `Service.Spec.ClusterIP` (not in `ClusterIPs`) and even if it's IPv6 address it tries to take it but then fails at IP validation:
```yaml
$ oc get svc httpserver-ipv6 -o yaml | grep -A 10 '^spec:'
spec:
  clusterIP: fd02::b244
  clusterIPs:
  - fd02::b244
  ipFamilies:
  - IPv6
  ipFamilyPolicy: SingleStack
  ports:
  - port: 8080
    protocol: TCP
    targetPort: 8080
```
```
time="2021-09-17T16:03:58Z" level=debug msg="Endpoints generated from service: publish-external-dns/httpserver-ipv6: [httpserver-ipv6.extdnstest.alebedev.com 0 IN A  fd02::b244 []]"
...
time="2021-09-17T16:05:00Z" level=error msg="InvalidChangeBatch: [Invalid Resource Record: 'FATAL problem: ARRDATAIllegalIPv4Address (Value is not a valid IPv4 address) encountered with 'fd02::b244'']\n\tstatus code: 400, request id: c762120b-aae3-42c3-b9bb-99418f7861b1"
```

This was already reported by [this upstream issue](https://github.com/kubernetes-sigs/external-dns/issues/1812).

File: pkg/operator/controller/externaldns/rbac.go
Comment (alebedev87) [reply pkg/operator/controller/externaldns/rbac.go:105]: No, they are not used for the moment ([my comment about this](https://github.com/openshift/external-dns-operator/pull/20#discussion_r710283331)).

File: README.md
Comment (alebedev87) [reply README.md]: Removed `TODO` section

**Comment by alebedev87** [comment]
/hold 

Hold until the commits are squashed (in case I get a LGTM ;) )

File: pkg/operator/controller/externaldns/rbac.go
Comment (Miciah) [inline pkg/operator/controller/externaldns/rbac.go:222]: What if expected is a proper superset of current?  For example, should the following test case pass?  
```go
		{
			name: "current is missing a rule",
			inputCurrent: []rbacv1.PolicyRule{},
			inputExpected: []rbacv1.PolicyRule{
				{
					APIGroups: []string{"networking.k8s.io"},
					Resources: []string{"ingresses"},
					Verbs:     []string{"get", "list", "watch"},
				},
			},
			expectChanged: true,
		},
```

File: pkg/operator/controller/externaldns/rbac_test.go
Comment (Miciah) [inline pkg/operator/controller/externaldns/rbac_test.go:168]: I guess `expectedExist` is always going to be true, right?  Still, it would be good form to check for nil before dereferencing `gotRole`.

File: pkg/operator/controller/externaldns/rbac_test.go
Comment (Miciah) [inline pkg/operator/controller/externaldns/rbac_test.go:222]: This rule isn't even in the `inputExpected` value; is that intentional?  This is more than just "reordered".

File: pkg/operator/controller/externaldns/rbac_test.go
Comment (Miciah) [inline pkg/operator/controller/externaldns/rbac_test.go:570]: I think you want the following:
```suggestion
						Name:      controller.ExternalDNSResourceName(testExternalDNS),
						Namespace: "otherns",
					},
				},
			},
			inputExpected: &rbacv1.ClusterRoleBinding{
				ObjectMeta: metav1.ObjectMeta{
					Name: controller.ExternalDNSResourceName(testExternalDNS),
				},
				RoleRef: rbacv1.RoleRef{
					APIGroup: "rbac.authorization.k8s.io",
					Kind:     "ClusterRole",
					Name:     controller.ExternalDNSBaseName,
				},
				Subjects: []rbacv1.Subject{
					{
						Kind:      "ServiceAccount",
						Name:      controller.ExternalDNSResourceName(testExternalDNS),
						Namespace: testNamespace,
```
That is, expected subject name and namespace and current subject name are the usual values, but current subject namespace is something other.

File: pkg/operator/controller/externaldns/rbac_test.go
Comment (Miciah) [inline pkg/operator/controller/externaldns/rbac_test.go:608]: I think you want `testNamespace`, which is defined in this file, not `namespace`, which is defined in `deployment_test.go`:
```suggestion
						Namespace: testNamespace,
```

File: api/v1alpha1/externaldns_types.go
Comment (Miciah) [inline api/v1alpha1/externaldns_types.go:421]: We should document what the default (empty) value signifies.  Presumably that means no DNS record is created if the source resource doesn't specify a host name, right?

**Review by Miciah — COMMENTED** [review]
Overall looks good; just some small suggestions.

File: api/v1alpha1/externaldns_types.go
Comment (alebedev87) [reply api/v1alpha1/externaldns_types.go:420]: To the ExternalDNS' source which is a Kubernetes service or OpenShift route.
It was my initiative to add more explanation to this flag, no doc exists upstream for this.

Updated the description to make this clearer.

File: config/rbac/role.yaml
Comment (alebedev87) [reply config/rbac/role.yaml:46]: Good catch, yes indeed it should not be used staring from OCP 4.9 (Kube 1.22). Removed

File: pkg/operator/controller/externaldns/controller_test.go
Comment (alebedev87) [reply pkg/operator/controller/externaldns/controller_test.go:131]: The ownership link is covered in the follow up PR (https://github.com/openshift/external-dns-operator/pull/22 still WIP). So, for the moment no removal of the secondary resources is implemented.
Yes, I would have to add the case for the deletion too.

File: pkg/operator/controller/externaldns/controller_test.go
Comment (alebedev87) [reply pkg/operator/controller/externaldns/controller_test.go:144]: Don't know, I think it's a brain glitch :) Renamed.

File: pkg/operator/controller/externaldns/rbac.go
Comment (alebedev87) [reply pkg/operator/controller/externaldns/rbac.go:102]: Removed

File: pkg/operator/controller/externaldns/rbac_test.go
Comment (alebedev87) [reply pkg/operator/controller/externaldns/rbac_test.go:222]: The idea was that even if all is reordered the content is the same. I changed the test name to make it clearer. Also I removed the `extentions/ingress` resources as per your previous comments.

File: pkg/operator/controller/externaldns/rbac.go
Comment (alebedev87) [reply pkg/operator/controller/externaldns/rbac.go]: There was a bug in the comparison. I found it in the later PRs, Andy did find it during his review and you did too. We are all good :)
I changed the implementation of this function to get 2 maps with sorted values (slices) and I compare them using `go-cmp`.

File: pkg/operator/controller/externaldns/rbac_test.go
Comment (alebedev87) [reply pkg/operator/controller/externaldns/rbac_test.go:168]: Yes, but since it's a part of testCase it can be set to `false`, so better to protect the dereferencing. Changed.

File: pkg/operator/controller/externaldns/rbac.go
Comment (alebedev87) [reply pkg/operator/controller/externaldns/rbac.go:63]: Done in my cloned copy.

File: pkg/operator/controller/externaldns/rbac.go
Comment (alebedev87) [reply pkg/operator/controller/externaldns/rbac.go]: N/A anymore, removed after the refactoring of the function.

File: api/v1alpha1/externaldns_types.go
Comment (alebedev87) [reply api/v1alpha1/externaldns_types.go:421]: Currently: default value (empty string) means no `--fqdn-template` flag will be added which in combination with the default hostname annotation policy (Ignore) would result into externalDNS container failing in the loop.

A possibility which I considered was add `--fdqn-template` even if the value is empty this would result in the behavior which you described: no records will be populated because none would match the empty FQDNTemplate. 

But I thought that it would be better to let the user know that the configuration is not optimal. That's why I added a task in [CFE-114](https://issues.redhat.com/browse/CFE-114) to add the validation of FQDNTemplate if the hostname annotation policy is Ignore.

Updated the doc to state explicitly that the value should not be empty if the hostname policy is Ignore.

**Comment by frobware** [comment]
As we discussed offline:
- testing of updateExternalDNSDeployment will be done as a follow-up

In the meantime let's get wider exposure & some e2e testing in place:

/lgtm

**Review by frobware — APPROVED** [review]

File: config/rbac/role.yaml
Comment (alebedev87) [reply config/rbac/role.yaml:12]: Comment in Jira epic: [link](https://issues.redhat.com/browse/NE-303?focusedCommentId=19051901&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-19051901).

File: pkg/operator/controller/externaldns/rbac.go
Comment (Miciah) [inline pkg/operator/controller/externaldns/rbac.go:98]: For a follow-up:
```suggestion
			Name: controller.ExternalDNSGlobalResourceName(),
￼```

File: pkg/operator/controller/externaldns/rbac.go
Comment (Miciah) [reply pkg/operator/controller/externaldns/rbac.go]: If we need to ignore ordering, it might be worth using go-cmp's `SortSlices` transformer so that you don't need `buildSortedPolicyRuleMap`.

File: api/v1alpha1/externaldns_types.go
Comment (Miciah) [reply api/v1alpha1/externaldns_types.go:421]: Is it important to allow an empty value?

By the way, as this allows multiple values, it should be a slice of strings, and we should document what specifying multiple strings does (presumably, for a given service, external-dns creates one DNS record for each string).

Comment (alebedev87) [reply api/v1alpha1/externaldns_types.go:421]: Yes, true it should have been a slice, my bad. Specifying multiple strings would create 1 DNS record for every **matching** domain. And the domain is found from the list of zone IDs or from the domains filtered by the domain filter (in the future).

**Pull request merged**
